### ISSUE TITLE
Federation falling behind, with `Event %s failed history check` errors

### ISSUE THREAD

**Issue by richvdh** [matrix-org/synapse#3923]
@turt2live reported problems with his homeserver getting behind. Inspection of the logs initially revealed the suspicious log line `Event %s failed history check`

**Comment by richvdh** [comment]
That log line was really unhelpful, and #3904 made significant improvements, including turning that %s into an actual event id.

After Travis had been running #3904 for a while, we then saw:

```
homeserver_1 - 2018-09-19 07:13:12,585 - synapse.handlers.federation - 345 - WARNING - PUT-138932 - [!zCLWAbYiBPnmFazMZQ:t2l.io $153734000845WPtXS:intelfx.name] Error attempting to resolve state at missing prev_events
Traceback (most recent call last):
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/synapse/handlers/federation.py", line 339, in on_receive_pdu
    state = (yield self.store.get_events(state_map.values())).values()
TypeError: unhashable type: 'list'
```

This turns out to have been a bug introduced by #3456, where `on_receive_pdu` calls `get_state_groups` expecting it to return a list of state groups, but it actually returns a dict mapping from state_group id to state group.

**Comment by richvdh** [comment]
fixed by #3968
